File: hr_os_frontend/hr_os_frontend/src/api/payroll.ts
import api from "@/lib/api";

export interface PayrollRun {
  id: string;
  month: number;
  year: number;
  status: string;
  total_gross: number;
  total_deductions: number;
  total_net: number;
  created_at: string;
  processed_at?: string;
}

export interface PayrollItem {
  id: string;
  employee_id: string;
  employee_name?: string;
  basic: number;
  hra: number;
  allowances: number;
  bonus: number;
  pf: number;
  esi: number;
  tax: number;
  leave_deduction: number;
  gross_salary: number;
  total_deductions: number;
  net_salary: number;
}

export const getPayrollRuns = async () => {
  const data = await api.get("/payroll/runs");
  return data as PayrollRun[];
};

export const runPayroll = async (month: number, year: number) => {
  const data = await api.post("/payroll/run", { month, year });
  return data as PayrollRun;
};

export const getPayrollDetail = async (id: string) => {
  const data = await api.get(`/payroll/runs/${id}`);
  return data as PayrollRun;
};

export const getPayslipDownloadUrl = (payrollId: string) => {
  const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || "http://127.0.0.1:8000";
  return `${apiBaseUrl}/api/v1/payroll/${payrollId}/payslip`;
};

File: hr_os_backend/app/schemas/payroll.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional
from datetime import date, datetime
import uuid

class EmployeeSalaryBase(BaseModel):
    employee_id: uuid.UUID
    ctc: float
    basic: float
    hra: float
    allowances: float
    bonus: float
    pf_employee: float
    pf_employer: float
    esi: float
    professional_tax: float

class EmployeeSalaryCreate(EmployeeSalaryBase):
    pass

class EmployeeSalaryOut(EmployeeSalaryBase):
    id: uuid.UUID
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class PayrollRunRequest(BaseModel):
    month: int
    year: int

class PayrollItemOut(BaseModel):
    id: uuid.UUID
    employee_id: uuid.UUID
    employee_name: Optional[str] = None
    basic: float
    hra: float
    allowances: float
    bonus: float
    pf: float
    esi: float
    tax: float
    leave_deduction: float
    gross_salary: float
    total_deductions: float
    net_salary: float
    working_days: float
    absent_days: float
    half_days: float
    insights: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)

class PayrollOut(BaseModel):
    id: uuid.UUID
    month: int
    year: int
    status: str
    total_gross: float
    total_deductions: float
    total_net: float
    created_at: datetime
    processed_at: Optional[datetime] = None
    items: List[PayrollItemOut] = []
    model_config = ConfigDict(from_attributes=True)
